crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2318,16 +2318,33 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_boolean_expression(&mut self, bool_op: &ast::ExprBoolOp) -> Type<'db> {
         let ast::ExprBoolOp {
             range: _,
-            op: _,
+            op,
             values,
         } = bool_op;
-
-        for value in values {
-            self.infer_expression(value);
+        let mut value_tys = Vec::new();
+        let inferred_types: Vec<Type<'db>> = values
+            .iter()
+            .map(|value| self.infer_expression(value))
+            .collect();
```

Comment:
Given that `UnionType::from_elements` takes an iterator, I think we can avoid all allocations here. It's slightly tricky because we do need to `infer_expression` on every value (even if we can short-circuit truthiness eval on some of them), but I think it is worth doing. Here's what I came up, which passes all tests (including the ones I suggested adding below):

<details>

```diff
diff --git a/crates/red_knot_python_semantic/src/types/infer.rs b/crates/red_knot_python_semantic/src/types/infer.rs
index 4decd441c..8f57befba 100644
--- a/crates/red_knot_python_semantic/src/types/infer.rs
+++ b/crates/red_knot_python_semantic/src/types/infer.rs
@@ -2321,30 +2321,36 @@ impl<'db> TypeInferenceBuilder<'db> {
             op,
             values,
         } = bool_op;
-        let mut value_tys = Vec::new();
-        let inferred_types: Vec<Type<'db>> = values
-            .iter()
-            .map(|value| self.infer_expression(value))
-            .collect();
-        for (i, value_type) in inferred_types.iter().enumerate() {
-            let boolean_value = value_type.bool(self.db);
-            if let Some(boolean_value) = boolean_value {
-                let is_last = i == values.len() - 1;
-                match (boolean_value, is_last, op) {
-                    (true, false, ast::BoolOp::And) => continue,
-                    (false, _, ast::BoolOp::And) => return *value_type,
-                    (true, _, ast::BoolOp::Or) => {
-                        value_tys.push(value_type);
-                        break;
+        let mut done = false;
+        UnionType::from_elements(
+            self.db,
+            values.iter().enumerate().map(|(i, value)| {
+                let value_ty = self.infer_expression(value);
+                if done {
+                    Type::Never
+                } else {
+                    let is_truthy = value_ty.bool(self.db);
+                    if let Some(boolean_value) = is_truthy {
+                        let is_last = i == values.len() - 1;
+                        match (boolean_value, is_last, op) {
+                            (true, false, ast::BoolOp::And) => Type::Never,
+                            (false, _, ast::BoolOp::And) => {
+                                done = true;
+                                value_ty
+                            }
+                            (true, _, ast::BoolOp::Or) => {
+                                done = true;
+                                value_ty
+                            }
+                            (false, false, ast::BoolOp::Or) => Type::Never,
+                            (_, true, _) => value_ty,
+                        }
+                    } else {
+                        value_ty
                     }
-                    (false, false, ast::BoolOp::Or) => continue,
-                    (_, true, _) => value_tys.push(value_type),
                 }
-            } else {
-                value_tys.push(value_type);
-            }
-        }
-        UnionType::from_elements(self.db, value_tys)
+            }),
+        )
     }
```

</details>